File: apple/Sources/ReactNativeWatchOS/Modifiers/Transition.swift
import SwiftUI

/// SwiftUI `.transition(_:)`. Maps the JS `type` preset to an
/// `AnyTransition`; `move` additionally reads an `edge` ('top'/'bottom'/
/// 'leading'/'trailing', defaulting to `.bottom`).
enum RNWTransitionModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("transition") { view, params, _ in
            let type = params.string("type") ?? "opacity"
            return AnyView(view.transition(make(type: type, edge: params.string("edge"))))
        }
    }

    private static func make(type: String, edge: String?) -> AnyTransition {
        switch type {
        case "slide":    return .slide
        case "scale":    return .scale
        case "identity": return .identity
        case "move":     return .move(edge: parseEdge(edge))
        default:         return .opacity
        }
    }

    private static func parseEdge(_ s: String?) -> Edge {
        switch s {
        case "top":      return .top
        case "leading":  return .leading
        case "trailing": return .trailing
        default:         return .bottom
        }
    }
}

File: src/modifiers/transition.ts
import { createModifier } from './createModifier';

export type TransitionType =
  | 'opacity'
  | 'slide'
  | 'scale'
  | 'identity'
  | 'move';

export type TransitionEdge = 'top' | 'bottom' | 'leading' | 'trailing';

export interface TransitionParams {
  /// The transition preset. Defaults to `'opacity'`.
  type?: TransitionType;
  /// Edge to move from/to. Only honored when `type` is `'move'`
  /// (defaults to `'bottom'`).
  edge?: TransitionEdge;
}

/// SwiftUI `.transition(_:)`. Describes how the view animates in/out when it
/// is inserted into or removed from the view tree (pair with `animation`).
///
/// ```tsx
/// <Text modifiers={[transition({ type: 'move', edge: 'leading' })]} />
/// ```
export function transition(
  type?: TransitionType
): ReturnType<typeof createModifier>;
export function transition(
  params: TransitionParams
): ReturnType<typeof createModifier>;
export function transition(a: TransitionType | TransitionParams = 'opacity') {
  if (typeof a === 'string') return createModifier('transition', { type: a });
  return createModifier('transition', a);
}
